instance/tests/management/test_instance_statistics_csv.py

```diff
@@ -0,0 +1,252 @@
+        """
+        Verify that today works
+        """
+        self.assertEqual(valid_date('2020-01-01'), '2020-01-01')
+
+    @freezegun.freeze_time('2020-01-01')
+    def test_date_in_past(self):
+        """
+        Verify that a date in the past works
+        """
+        self.assertEqual(valid_date('2019-01-01'), '2019-01-01')
+
+
+class InstanceStatisticsCSVTestCase(TestCase):
+    """
+    Test cases for the `instance_statistics_csv` management command.
+    """
+    def test_no_qualified_domain(self):
+        """
+        Verify that the command correctly notifies the user that the domain is required
+        """
+        with self.assertRaises(CommandError) as context_manager:
+            call_command(
+                'instance_statistics_csv'
+            )
+            self.assertIn('Error: the following arguments are required', str(context_manager.exception))
+
+    @patch('instance.management.commands.instance_statistics_csv.open', Mock(side_effect=PermissionError))
+    def test_unable_to_write_to_output(self):
+        """
+        Verify that the command correctly notifies the user when it is unable to write to the request output file
+        """
+        stderr = StringIO()
+        with self.assertRaises(SystemExit):
+            call_command(
+                'instance_statistics_csv',
+                domain='test',
+                out='does_not_exist.txt',
+                stderr=stderr
+            )
+        self.assertIn('Permission denied while attempting to write file', stderr.getvalue())
+
+    def test_no_matching_instances(self):
+        """
+        Verify that the command correctly notifies the user that
+        there are no matching instances.
+        """
+        stderr = StringIO()
+        with self.assertRaises(SystemExit):
+            call_command(
+                'instance_statistics_csv',
+                domain='test.opencraft.hosting',
+                stderr=stderr
+            )
+        self.assertIn('No OpenEdXInstance exists with an external or internal domain', stderr.getvalue())
+
+    @patch('instance.models.openedx_instance.OpenEdXInstance.objects')
+    def test_instance_not_successfully_provisioned(self, mock_openedx_instance):
+        """
+        Verify that the command correctly notifies the user that
+        there are no active app servers if the instance failed to provision.
+        """
+        mock_instance = PropertyMock(successfully_provisioned=False)
+        mock_openedx_instance.get.return_value = mock_instance
+
+        stderr = StringIO()
+        with self.assertRaises(SystemExit):
+            call_command(
+                'instance_statistics_csv',
+                domain='test.opencraft.hosting',
+                stderr=stderr
+            )
+        self.assertIn(
+            'No active OpenEdXAppServers exist for the instance',
+            stderr.getvalue()
+        )
+
+    @patch('instance.models.openedx_instance.OpenEdXInstance.objects')
```

Comment:
@pcockwell, I think the same comment for the previous unit test applies to this one as well. Even if you spawn an appserver, it is not activated by default. So there will be no active appservers.

Reply:
Done in next commit